dd-sdk-android-ndk/src/main/cpp/datadog-native-lib.cpp

```diff
@@ -0,0 +1,111 @@
+#include "datadog-native-lib.h"
+
+#include <cstring>
+#include <fstream>
+#include <jni.h>
+#include <pthread.h>
+#include <sstream>
+
+#include "android/log.h"
+#include "utils/backtrace-handler.h"
+#include "utils/datetime.h"
+#include "utils/fileutils.h"
+#include "utils/signal-monitor.h"
+
+static const char *GENERIC_MESSAGE = "Native crash detected";
+static const char *EMERGENCY_STATUS = "emergency";
+static const char *ENVIRONMENT = "";
+static const char *ERROR_KIND = "Native";
+static const char *LOGGER_NAME = "crash";
+static const char *LOG_TAG = "NdkCrashReporter";
+static const char *SERVICE_NAME = "";
+static const char *STORAGE_DIR = nullptr;
+
+std::string get_serialized_log(int signal,
+                               const char *signal_name,
+                               const char *error_message,
+                               const char *date,
+                               const char *backtrace) {
+    std::string serializedLog;
+    serializedLog.append("{ ");
+    serializedLog.append(R"("message": ")").append(GENERIC_MESSAGE).append("\",");
+    serializedLog.append(R"("service": ")").append(SERVICE_NAME).append("\",");
+    serializedLog.append(R"("logger.name": ")").append(LOGGER_NAME).append("\",");
+    serializedLog.append(R"("env": ")").append(ENVIRONMENT).append("\",");
+    serializedLog.append(R"("status": ")").append(EMERGENCY_STATUS).append("\",");
+    serializedLog.append(R"("date": ")").append(date).append("\",");
+    serializedLog.append(R"("error.stack": ")").append(backtrace).append("\",");
+    serializedLog.append(R"("error.message": ")").append(error_message).append("\",");
+    char formattedSignalMessage[30];
+    sprintf(formattedSignalMessage, "%s: %d", signal_name, signal);
+    serializedLog.append(R"("error.signal": ")").append(formattedSignalMessage).append("\",");
+    serializedLog.append(R"("error.kind": ")").append(ERROR_KIND).append("\"");
+    serializedLog.append(" }");
+    return serializedLog;
+}
+
+void crash_signal_intercepted(int signal, const char *signal_name, const char *error_message) {
+    // sync everything
+    static pthread_mutex_t handlerMutex = PTHREAD_MUTEX_INITIALIZER;
+    pthread_mutex_lock(&handlerMutex);
+
+    if (STORAGE_DIR == nullptr) {
+        __android_log_print(ANDROID_LOG_ERROR, LOG_TAG,
+                            "%s",
+                            "The crash reports storage directory file path was null");
```

Comment:
You must unlock the mutex before returning

Reply:
oh...right is not like in Java :(